doc/explanation/mental-model.rst

```diff
@@ -0,0 +1,197 @@
+    it can be attached to *aliases*).
+
+When ``dune build`` is executed, it will first read the project's ``dune`` files to determine
+the rules that apply to the project. Once it has done this, it will determine
+what actions it needs to execute to build the required targets.
+
+An Example
+----------
+
+Let's take the following example.
+
+- there's a CLI tool written in OCaml.
+- it has some build-time configuration stored in ``config.json``.
+- it has an integration test, in which the tool is executed with
+  ``testdata.txt`` as input.
+
+Configuration Generation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+To express the generation of the configuration module we could write:
+
+.. code:: dune
+
+  (rule
+   (deps convert/json2ml.exe config.json)
+   (target config.ml)
+   (action
+    (run convert/json2ml.exe config.json -o config.ml)))
+
+This rule will:
+
+- read its dependencies: ``convert/json2ml.exe`` and ``config.json``
+- and write its target: ``config.ml``
+- using an action: ``(run convert/json2ml.exe config.json -o config.ml)``
+
+This rule is very explicit: we write a stanza for a single Dune rule.
+
+Building the Executable
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In contrast, to describe the compilation of the executable, we would write:
+
+.. code:: dune
+
+  (executable
+   (name tool)
+   (modules main config))
+
+Here, we use Dune's abstractions. Dune knows about the OCaml compilation model:
+the modules need to be compiled and linked together. So it will generate the
+following rules under the hood:
+
+- one rule to compile the ``Main`` module:
+
+  - it will read its dependency: ``main.ml``
+  - and write its output: ``main.cmx``
+  - using an action: ``(run ocamlopt -c main.ml)``
+
+- one rule to compile the ``Config`` module:
+
+  - it will read its dependency: ``config.ml``
+  - and write its output: ``config.cmx``
+  - using an action: ``(run ocamlopt -c config.ml)``
+
+- one rule to link the ``tool.exe`` executable:
+
+  - it will read its dependencies: ``main.cmx`` and ``config.cmx``
+  - and write its output: ``tool.exe``
+  - using an action: ``(run ocamlopt -o tool.exe main.cmx config.cmx``)
+
+Note that in this example, some files are targets of a rule and dependencies of
+another (``.cmx`` files). We are unlikely to ever interact with them directly,
+so it can also be useful to think of the ``(executable)`` stanza as a group of
+rules with ``main.ml`` and ``config.ml`` as inputs and ``tool.exe`` as output.
+
+Running the Tests
+^^^^^^^^^^^^^^^^^
+
+Some rules do not produce any output file, but we're still interested in
+running their actions. A test is a good example: we want the build process to exit
+with an error code if the action fails.
+In that case, the rule does not have targets, but we "attach" it to an
+:term:`alias`, ``runtest`` in this case. This gives us a way of requesting this
+rule to be executed. As we are about to see, rules are executed lazily by
```

Comment:
While this sort of implies it is possible I think it would be nice to state directly that more than one rule can be attached to an alias and in fact `runtest` has lots of rules attached to them.

(That's one way my understanding of aliases as "phony Makefile targets" did not match up)

Reply:
a "dune for make users" guide could be useful indeed!